### DISCUSSION TITLE
Opting out of -- token parsing to improve interop with called programs

### DISCUSSION THREAD

**Discussion by cyanite** [PowerShell/PowerShell#18143] [Enhancements]
PowerShell does a lot of processing on behalf of its cmdlets, advanced functions etc. Sometimes this is a problem when you want to make "advanced aliases" (which in PowerShell will have to be functions). For example:

    function ggl { git log graph @args }
    # but could be more complex, e.g. inspect the arguments and auto-add some options.

The problem is that -- means something in Git, beyond just terminating option parsing: It separates revision arguments from file arguments. So I'd like to say,

    ggl some-branch -- Some/Path

And this might work... but git never sees the -- token, so it might also not work, if there is any ambiguity. There is no way to have PowerShell not process this token so it can be passed on to sub-programs.

I deliberately make "ggl" a non-advanced function, since otherwise PowerShell would interfere even more with parsing.

**Comment by jborean93** [comment]
This seems to work as normal for me

```powershell
PS C:\Users\vagrant-domain> C:\temp\print_argv.exe arg1 -- test
"C:\temp\print_argv.exe" arg1 -- test
[1] arg1
[2] --
[3] test
```

The `print_argv,.exe` is a custom binary that outputs the full command line + how it parsed it into the separate arguments. You can see how `--` was seen as normal. You can also quote the value so PowerShell just sees it as a string but that shouldn't matter at all in this case.

**Reply by cyanite** [reply]
Yes, it does work for foreign binaries. But my use case is a function, acting as an alias. And as soon as it's a PowerShell cmdlet or function, the "--" processing is not possible to avoid.

Yes, you can quote it, but that moves the burden to the user of the function. Now, every time you call it you must type

    ggl something `-- somefile

**Reply by jborean93** [reply]
Ah my apologies I misread your question. I'm not sure you can get around this without quoting as `--` when in parsing mode does the binder to treat the rest of the parameters as bare strings. You can prefix it with another `--` if you wish but I'm not sure how nice that is compared to just quoting the `--`, e.g. `ggl -- something -- somefile`.

**Comment by herronelou** [comment]
Hello.
I'm reviving this thread to see if an alternative was ever found, I'm in the nearly exact same case as @cyanite, although with a different context (need to replicate the functionality of bash aliases as part of a tool that handles different shells). 
Currently using functions as shown in the original post, but it does break when some arguments get parsed by powershell before reaching the function (for example `--`).

I went down a rabbit hole of trying to obtain the raw command and re-parsing that, but it feels dangerous and convoluted.
